### DISCUSSION TITLE
Are there any plans to offer an API?

### DISCUSSION THREAD

**Discussion by gtusr** [florisboard/florisboard#2070] [Feedback]
I imagine many users would like to customize certain aspects of this keyboard by building their own components ... not just by adjusting settings.  Tasker comes to mind as a tool that some might want to use to extend their keyboard's feature set.

Fleksy already offers an [SDK](https://www.fleksy.com/) apparently ... but not sure if there are any others out there.

**Comment by patrickgold** [comment]
FlorisBoard has been built and designed as an end product for users from the beginning, but not really as an SDK where developers can create their own keyboards, as this project's target user group are end users. Doesn't mean FlorisBoard is completely static, themes, layouts and dictionaries are pretty modular internally, although only themes have a somewhat good frontend at the moment. There are plans for even more customization options (especially everything that affects themes, sounds and layouts), but that's still far into the future.

As you mentioned Tasker, there have been requests about support for Tasker or other automation tools and I am pretty open to this, but there are a lot of different areas that require more attention atm so Tasker is pretty low on my priority list.

**Comment by gtusr** [comment]
Fair enough.  

The use case of greatest interest to me at the moment is to configure a better keyboard for use with emacs in termux, which I know is quite specialized.  

Maybe SDK was too strong a word ... still curious though if there are official plans to encourage development of 'keyboard extensions' through formalizing an API analogous to WebExtensions for browsers? 
 
Making it easy for example to plug in different components for word prediction and 'swype' recognition could be of benefit to some.

I also have an interest in exploring (much further down my to-do list) to what extent alternative keyboard layouts, that could also include alternative geometries (e.g., a hexagonal grid of keys) could be useful for improving speed and accuracy when 'swyping.'

The wonderful thing about open source is that anyone can download the source and modify it.  Gboard is a fantastic keyboard ... but unfortunately Google has made a decision to hold back that technology from AOSP Android.

So thanks for all of your work to bring this project as far as it has already come!  I've subscribed to the latest alpha channel and will be eagerly testing it from time to time in order to determine when I switch over to Florisboard for my daily driver.

**Comment by patrickgold** [comment]
> still curious though if there are official plans to encourage development of 'keyboard extensions' through formalizing an API analogous to WebExtensions for browsers?

Yes there are and that project is called "FlorisBoard extensions". Themes and internally layouts already make use of this format. Extension files have the extension `.flex`, but they are basically just ZIP files. In there an `extension.json` must be defined, which provides meta data and other configs. This format is however not stable yet and subject to change in the future (especially in the 0.4 and 0.5 releases). Also Flex extensions are much less powerful than web extensions - they only allow static json and other files, no scripting languages.

> I also have an interest in exploring (much further down my to-do list) to what extent alternative keyboard layouts, that could also include alternative geometries (e.g., a hexagonal grid of keys) could be useful for improving speed and accuracy when 'swyping.'

Not possible atm but certainly not impossible either in the future. Just needs a whole rework of the text keyboard logic and keyboard extension format.
